riemann_c_client: Optional TLS support & other changes

algernon opened the pull request:
This is a small update to riemann_c_client to optionally enable features, and correct a few things in the packaging.

###### Description of changes

- Allow building with wolfSSL or GnuTLS (both disabled by default), and with json_c (enabled by default, but can be turned off)
- Split outputs to `bin`, `dev`, and `out`
- Enable parallel building
- Remove the preBuild step, the bug it was working around has been fixed upstream.
- Fix the license: it is LGPL3+, not GPL3.
- Enable the unit tests during build.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

pbsds commented on pkgs/tools/misc/riemann-c-client/default.nix:
```diff
@@ -1,27 +1,39 @@
-  src = fetchFromGitHub {
-    owner = "algernon";
-    repo = "riemann-c-client";
+  src = fetchgit {
```
Please consider using this one:

```suggestion
  src = fetchFromGitea {
```

pbsds commented on pkgs/tools/misc/riemann-c-client/default.nix:
```diff
@@ -1,27 +1,39 @@
+    url = "https://git.madhouse-project.org/algernon/riemann-c-client.git";
     rev = "riemann-c-client-${version}";
-    sha256 = "sha256-LuI9XFDPx0qw/+kkpXd0FOMESERAp31R1+ttkGuJnPA=";
+    sha256 = "sha256-UaB/iMJh1t0K99FRwmMkRci8juKPro9vf8pNlEao5Ho=";
```
```suggestion
    hash = "sha256-UaB/iMJh1t0K99FRwmMkRci8juKPro9vf8pNlEao5Ho=";
```

pbsds reviewed (changes requested):
Looking good! Seems like this PR could use a rebase onto master

pbsds commented on pkgs/tools/misc/riemann-c-client/default.nix:
```diff
@@ -9,20 +13,29 @@ stdenv.mkDerivation rec {
+  buildInputs = [ file protobufc ]
+    ++ lib.optional withWolfSSL wolfssl
+    ++ lib.optional withGnuTLS gnutls
+    ++ lib.optional withJSON json_c
```
Is json_c picked up by the config script automatically? Is there a reason to not include it by default?

pbsds commented:
Unit tests are very much welcome. Integration testing with the server on the other hand is better relegated to a `passthru.tests` test, if included at all.

I'd say include the json dep. :)

algernon committed:
riemann_c_client: Optional TLS support & other changes

This adjusts riemann_c_client to support building with either wolfSSL or
GnuTLS (disabled by default, to not introduce a new dependency
automatically), and the CLI tool with json-c (enabled by default).

Also splits the outputs to `bin`, `dev`, and `out`, enables parallel
building, and removes a preBuild step that is no longer necessary, and
enables unit tests at build time.

On top of this, also correct the license: it is LGPL3+, *not* GPL3.

Signed-off-by: Gergely Nagy <me@gergo.csillger.hu>

pbsds commented:
Result of `nixpkgs-review pr 243131` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li><code>collectd</code></li>
    <li><code>riemann_c_client</code></li>
    <li><code>riemann_c_client.bin</code></li>
    <li><code>riemann_c_client.dev</code></li>
    <li><code>riemann_c_client.override (old: { withGnuTLS = true; })</code></li>
    <li><code>riemann_c_client.override (old: { withWolfSSL = true; })</code></li>
    <li><code>syslogng</code></li>
    <li><code>syslogng.man</code></li>
  </ul>
</details>

---

- [x] package version fits guidelines
- [x] package build on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] all depending packages build
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1100

algernon commented:
Anything I can do to help push this forward?

pbsds reviewed (commented):
I've recieved merge access since i reviewed this. Rebuilding on current master...

Result of `nixpkgs-review pr 243131` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>collectd</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>riemann_c_client</li>
    <li>riemann_c_client.bin</li>
    <li>riemann_c_client.dev</li>
    <li>syslogng</li>
    <li>syslogng.man</li>
  </ul>
</details>


The `collectd` failure [also happens on master](https://hydra.nixos.org/build/238910489), with the same error.

The following also build:

* `riemann_c_client.override { withWolfSSL = true; })`
* `riemann_c_client.override { withGnuTLS = true; })`
* `riemann_c_client.override { withJSON = false; })`
